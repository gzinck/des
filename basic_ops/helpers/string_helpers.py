from json import dumps


def get_states(states, chosen_so_far=[]):
    """A list of current states for multiple automata in a composed system (i.e.,
    a "macro-state") has some states which are non-deterministic: that is, the
    state might be ["q1", ["q2", "q3"]]. We want to find all distinct states,
    so we want ["q1", "q2"] and ["q1", "q3"]. This becomes a major task when
    there are many options for many automata, leading to a combinatorial
    explosion.

    Parameters
    ----------
    states : array of strings, or array of arrays of strings
        Array of all states (or possible states)

    Returns
    -------
    Array of arrays of strings
        All possible macro-states

    Examples
    --------
    >>> print(get_states(["q1", ["q2", "q3"]]))
    [["q1", "q2"], ["q1", "q3"]]
    """
    index = len(chosen_so_far)

    # Base case
    if index == len(states):
        return [chosen_so_far.copy()]

    macro_states = []
    for option in states[index]:
        chosen_so_far.append(option)
        macro_states.extend(get_states(states, chosen_so_far))
        chosen_so_far.pop()
    return macro_states


def format_state(states):
    """Creates a macro-state string containing all of the states passed in.

    Parameters
    ----------
    states : array of strings
        Array of all the states to put together

    Returns
    -------
    string
        The string representing the macro-state

    Examples
    --------
    >>> print(format_state(["q1"]))
    "(q1)"
    >>> print(format_state(["q1", "q2", "(q3, q4)"]))
    "(q1, q2, (q3, q4))"
    """
    str = "("
    for state in states:
        str += state + ", "
    str = str[0:-2]  # take off last comma
    str += ")"
    return str


def format_state_set(states):
    """Creates a macro-state string containing all of the states passed in as a
    set.

    Parameters
    ----------
    states : array of strings
        Array of all the states to put together

    Returns
    -------
    string
        The string representing the macro-state

    Examples
    --------
    >>> print(format_state(["q1"]))
    "{q1}"
    >>> print(format_state(["q1", "q2", "(q3, q4)"]))
    "{q1, q2, (q3, q4)}"
    """
    states.sort()
    str = "{"
    for state in states:
        str += state + ", "
    str = str[0:-2]  # take off last comma
    str += "}"
    return str


def format_event_vector(events):
    """Formats an event vector into a format with angle brackets.

    Parameters
    ----------
    events : list
        List of strings representing the events to put together

    Returns
    -------
    str
        The string representing the event
    """
    s = "["
    for e in events:
        s += e + ", "
    s = s[:-2]
    s += "]"
    return s


def format_transition(state, event):
    """Formats a state and event into the proper format for a transition,
    which is used as a key in the transition dictionary.

    Parameters
    ----------
    state : string
        The state the transition originates from
    event : string
        The event prompting the state change

    Returns
    -------
    string
        The string representing the transition's key for the dictionary

    Examples
    --------
    >>> print(format_transition("q1", "a"))
    "q1->a"
    >>> print(format_transition("(q1, q5)", "c"))
    "(q1, q5)->c"
    """
    return state + "->" + event


def format_observed_secrets(observer, observed_secrets):
    """Formats an observer's observation of other agents' secrets

    Parameters
    ----------
    observer : str
        The observer that has identified secrets
    observed_secrets : list
        List of identified secrets

    Returns
    -------
    str
        The string representing the agent's observations
    """
    lst = ""
    for x in observed_secrets:
        lst += ", " + str(x)
    if len(lst) > 2:
        lst = lst[2:]
    return str(observer) + "->" + "{" + lst + "}"


def format_all_observed_secrets(lst):
    """Formats all observers' observations of other agents' secrets

    Parameters
    ----------
    lst : list
        List of observations (see format_observed_secrets).

    Returns
    -------
    str
        The string representing all agents' observations
    """
    secrets = [format_observed_secrets(*x) for x in lst]
    secrets.sort()
    result = ""
    for s in secrets:
        result += ", " + s
    if len(result) > 2:
        result = result[2:]
    return result


def extract_state(transition):
    """Extracts the name of the origin state given a string representing a
    transition.

    Parameters
    ----------
    transition : str
        A transition

    Returns
    -------
    str
        The corresponding state from the transition
    """
    return transition.split("->", 1)[0]


def extract_event(transition):
    """Extracts the name of the event given a string representing a
    transition.

    Parameters
    ----------
    transition : str
        A transition

    Returns
    -------
    str
        The corresponding event from the transition
    """
    return transition.split("->", 1)[1]


def set_since_tau(state, events_since_tau):
    """Sets the number of events since tau in the name of the state.

    Parameters
    ----------
    state: str
    events_since_tau: int
        The number of events that have occurred since τ occurred

    Returns
    -------
    str
        The new state name that incorporates the number of events since τ.
    """
    return f'{state}_{events_since_tau}'


def extract_since_tau(state):
    """Extracts the number of events that have occurred since an event τ.
    Only works if state is formatted as "5_0", representing 0 events since
    τ.

    Parameters
    ----------
    state: str

    Returns
    -------
    int
        The number of events since the tau occurred
    """
    return int(state.rsplit("_", 1)[1])


def extract_state_without_since_tau(state):
    """Extracts the name of the state without the number of events that have
    occurred since an event τ. Only works if state is formatted as "5_0",
    representing 0 events since τ at state 5.

    Parameters
    ----------
    state: str

    Returns
    -------
    str
        The state without the since tau
    """
    return state.rsplit("_", 1)[0]


def pretty_print(automaton):
    """
    Formats an automaton in an attractive way and prints as text.

    Parameters
    ----------
    automaton : dict
        The automaton to print

    Returns
    -------
    None
    """
    print(dumps(automaton, sort_keys=True, indent=4))
