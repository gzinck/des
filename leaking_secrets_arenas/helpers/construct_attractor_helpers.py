def mark_bad_v2s(attractor):
    """Adds bad-v2 states to the attractor—that is, v2 states which lead to
    bad-v1 states.

    Parameters
    ----------
    attractor : dictionary
        The attractor, which is essentially an arena but with more bad states

    Returns
    -------
    boolean
        Whether or not there were any extra bad-v2 added
    """
    bad_added = False
    all_trans = attractor["transitions"]["v2"]
    for v2_state in attractor["states"]["v2"]:

        # Don't worry about this state if already bad
        if v2_state in attractor["states"]["bad-v2"]:
            continue

        nxt_options = [v for k, v in all_trans.items() if v2_state in k]

        # If there are no transitions, we're fine
        if len(nxt_options) == 0:
            continue

        # Go through every transition, if one leads to bad, we're not ok
        is_bad = False
        for nxt_set in nxt_options:
            for nxt in nxt_set:
                if nxt in attractor["states"]["bad-v1"]:
                    is_bad = True

        # If no good options, this is a bad state
        if is_bad:
            attractor["states"]["bad-v2"].append(v2_state)
            bad_added = True
    return bad_added


def mark_bad_v1s(attractor):
    """Adds bad-v1 states to the attractor—that is, v1 states for which all
    control decisions lead to bad-v2 states.

    Parameters
    ----------
    attractor : dictionary
        The attractor, which is essentially an arena but with more bad states

    Returns
    -------
    boolean
        Whether or not there were any extra bad-v1s added
    """
    bad_added = False
    all_trans = attractor["transitions"]["v1"]
    for v1_state in attractor["states"]["v1"]:

        # Don't worry about this state if already bad
        if v1_state in attractor["states"]["bad-v1"]:
            continue

        nxt_options = [v for k, v in all_trans.items() if v1_state in k]

        # If there are no transitions, we're fine
        if len(nxt_options) == 0:
            continue

        # Go through every transition, if one does not lead to bad, we're happy
        all_bad = True
        for nxt_set in nxt_options:
            for nxt in nxt_set:
                if nxt not in attractor["states"]["bad-v2"]:
                    all_bad = False

        # If no good options, this is a bad state
        if all_bad:
            attractor["states"]["bad-v1"].append(v1_state)
            bad_added = True
    return bad_added
