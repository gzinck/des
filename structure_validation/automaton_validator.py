def validate(automaton):
    """Validates if an automaton is correctly formatted. This includes:

    1. Has a states area with "all", "initial", "marked", and "bad"
            - Must be a dictionary
            - "all", "initial", "marked", and "bad" are lists of strings
    2. Has an events area with "all", "controllable", "observable", and
       "attacker"
            - Must be a dictionary
            - "all" and "attacker" are lists of strings
            - "controllable" is a list of lists of strings, as is "observable"
              The two must have the same number of lists.
    3. Has a transitions area with "all" and "bad"
            - "all" and "bad" are dictionaries mapping strings to lists of
              strings

    If successful, returns True. If unsucessful, raises an ValueError explaining
    the problem.

    Parameters
    ----------
    automaton : dictionary
        The automaton to verify its structure

    Examples
    --------
    >>> validate(automaton1)
    True
    """

    if not isinstance(automaton, dict):
        raise ValueError("The automaton must be a dictionary structure")

    for x in ["states", "events", "transitions"]:
        if x not in automaton:
            raise ValueError("The automaton must have a key for " + x)

    # 1: STATES
    states = automaton["states"]
    if not isinstance(states, dict):
        raise ValueError("The states subtree must be a dictionary structure")

    for x in ["all", "initial", "bad"]:
        if x not in states:
            raise ValueError("The states subtree must have a key for " + x)
        for state in states[x]:
            if not isinstance(state, str):
                raise ValueError("The only permitted type inside the states subtree " + x + " is a string")

    # marked is separate, because it's a list of lists
    marked = states["marked"]
    if not isinstance(marked, list):
        raise ValueError("The marked states should be a list of lists")

    for x in marked:
        if not isinstance(marked, list):
            raise ValueError("The marked states should be a list of lists")
        for state in x:
            if not isinstance(state, str):
                raise ValueError("The only permitted type inside the marked states list is a string")

    # 2: EVENTS
    events = automaton["events"]
    if not isinstance(events, dict):
        raise ValueError("The events subtree must be a dictionary structure")

    for x in ["all", "controllable", "observable"]:
        if x not in events:
            raise ValueError("The events subtree must have a key for " + x)

    for x in ["all"]:
        for event in events[x]:
            if not isinstance(event, str):
                raise ValueError("The only permitted type inside the events subtree " + x + " is a string")

    for x in ["observable", "controllable"]:
        if not isinstance(events[x], list):
            raise ValueError("The " + observable + " area must be a list-of-lists-of-strings structure")

        for lst in events[x]:
            if not isinstance(lst, list):
                raise ValueError("The " + observable + " area must be a list-of-lists-of-strings structure")
            for event in lst:
                if not isinstance(event, str):
                    raise ValueError("The only permitted type inside the " + x + " list-of-lists is a string")

    if len(events["observable"]) != len(events["controllable"]):
        raise ValueError(
            "There must be the same number of lists within both observable and controllable (one for each player in the system)")

    # 3: Transitions
    trans = automaton["transitions"]
    if not isinstance(trans, dict):
        raise ValueError("The transitions subtree must be a dictionary structure")

    for x in ["all", "bad"]:
        if x not in trans:
            raise ValueError("The transitions subtree must have a key for " + x)
        # Make sure it's a dict
        if not isinstance(trans[x], dict):
            raise ValueError("The transitions subtree " + x + " must be a dictionary structure")
        # Make sure all entries are lists
        for key, value in trans[x].items():
            if not isinstance(value, list):
                raise ValueError("The values for transitions must be lists of strings")
            for state in value:
                if not isinstance(state, str):
                    raise ValueError("The values for transitions must be lists of strings")
    return True
