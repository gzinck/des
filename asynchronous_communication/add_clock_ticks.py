from basic_ops.helpers.string_helpers import *


def add_clock_ticks(automaton, delta):
    """Adds clock ticks every delta events to the automaton, starting with
    adding a clock tick at the beginning.

    Parameters
    ----------
    automaton: dict
        The automaton to add clock ticks to
    delta: int
        The frequency of clock ticks, where delta is the number of events
        between ticks
    """
    all_events = automaton["events"]["all"]

    # First, add a clock tick at the beginning
    old_initial = automaton["states"]["initial"][0]

    # Give the new initial state a new name
    new_initial = "0"

    old_trans = automaton["transitions"]["all"]
    old_spec = automaton["transitions"]["specification"]
    new_trans = {}
    new_spec = {}
    new_states = set()

    # Step 1: put a transition from new initial to the old one

    new_trans[format_transition(new_initial, "tau")] = [set_since_tau(old_initial, 0)]

    # Add states
    new_states.add(new_initial)
    new_states.add(set_since_tau(old_initial, 0))

    # Step 2: add taus every delta events
    # We implement with BFS
    queue = [set_since_tau(old_initial, 0)]

    while len(queue) > 0:
        curr = queue.pop(0)
        curr_state = extract_state_without_since_tau(curr)
        curr_taucount = extract_since_tau(curr)

        if curr_taucount < delta:
            # Go through every event
            for event in all_events:
                # Check if transition exists
                trans = format_transition(curr_state, event)
                if trans in old_trans:
                    # Should add transition
                    nxt_state = set_since_tau(old_trans[trans][0], curr_taucount + 1)
                    new_trans[format_transition(curr, event)] = [nxt_state]

                    # Add to specification if needed
                    if trans in old_spec:
                        new_spec[format_transition(curr, event)] = [nxt_state]

                    # Add the state
                    if nxt_state not in new_states:
                        new_states.add(nxt_state)
                        queue.append(nxt_state)

        else:
            # If we need to have a tau, tau it up!
            nxt_state = set_since_tau(curr_state, 0)
            new_trans[format_transition(curr, "tau")] = [nxt_state]
            new_spec[format_transition(curr, "tau")] = [nxt_state]
            if nxt_state not in new_states:
                new_states.add(nxt_state)
                queue.append(nxt_state)

    return {
        "events": {
            "all": all_events + ["tau"],
            "controllable": automaton["events"]["controllable"],
            "observable": automaton["events"]["observable"]
        },
        "states": {
            "all": list(new_states),
            "initial": [new_initial],
            "marked": automaton["states"]["marked"],
            "bad": automaton["states"]["bad"],
        },
        "transitions": {
            "all": new_trans,
            "specification": new_spec,
            "bad": automaton["transitions"]["bad"]
        }
    }
