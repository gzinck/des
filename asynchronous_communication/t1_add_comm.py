from basic_ops.helpers.string_helpers import *


def add_communication(automaton, agent):
    # Events
    reg_events = automaton["events"]["all"]
    rec_events = [f'?{event}' for event in reg_events if event != "tau"]
    send_events = [f'!{event}' for event in reg_events if event != "tau"]

    # Transitions
    old_trans = automaton["transitions"]["all"]
    old_spec = automaton["transitions"]["specification"]
    new_trans = {}
    new_spec = {}

    # States
    new_states = set()
    new_states.add(automaton["states"]["initial"][0])

    # Start off the queue with the initial state
    # Doubled because the state has no annotations
    queue = [[automaton["states"]["initial"][0]] * 2]

    while len(queue) > 0:
        # The state and the state without any annotations
        curr, curr_no_annot = queue.pop(0)

        # Go through every event
        for event in reg_events:
            orig_trans = format_transition(curr_no_annot, event)
            if orig_trans in old_trans:
                nxt = old_trans[orig_trans][0]
                # Then add it to the new one
                trans = format_transition(curr, event)
                new_trans[trans] = [nxt]
                # Check if we need to do the same in the spec
                if orig_trans in old_spec:
                    new_spec[trans] = [nxt]

                # Add to the queue with the newly reachable state
                if nxt not in new_states:
                    queue.append([nxt, nxt])
                    new_states.add(nxt)
                # If observable, also add send
                if event in automaton["events"]["observable"][agent]:
                    new_t = format_transition(nxt, f'!{event}')
                    new_nxt = f'{nxt}!'
                    new_trans[new_t] = [new_nxt]
                    # Add to spec if we should
                    if orig_trans in old_spec:
                        new_spec[new_t] = [new_nxt]
                    # Add to the queue
                    if new_nxt not in new_states:
                        queue.append([new_nxt, nxt])
                        new_states.add(new_nxt)
                else:
                    new_t = format_transition(nxt, f'?{event}')
                    new_nxt = f'{nxt}?'
                    new_trans[new_t] = [new_nxt]
                    # Add to spec if we should
                    if orig_trans in old_spec:
                        new_spec[new_t] = [new_nxt]
                    # Add to the queue
                    if new_nxt not in new_states:
                        queue.append([new_nxt, nxt])
                        new_states.add(new_nxt)

    return {
        "events": {
            "all": reg_events + rec_events + send_events,
            "controllable": automaton["events"]["controllable"],
            "observable": automaton["events"]["observable"],
            "regular": reg_events,
            "receive": rec_events,
            "send": send_events
        },
        "states": {
            "all": list(new_states),
            "initial": automaton["states"]["initial"],
            "marked": automaton["states"]["marked"],
            "bad": automaton["states"]["bad"]
        },
        "transitions": {
            "all": new_trans,
            "specification": new_spec,
            "bad": automaton["transitions"]["bad"]
        }
    }


