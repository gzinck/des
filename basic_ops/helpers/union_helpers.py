import basic_ops.helpers.string_helpers as helper
import basic_ops.helpers.state_helpers as state_helper


def union_events(automata):
    """Unions all of the events of multiple automata

    Parameters
    ----------
    automata : array of dictionaries
        Array of all of the automata which should be composed

    Returns
    -------
    dict
        Dictionary containing the union for all events, controllable events,
        and observable events.

    Examples
    --------
    >>> events = union_events([automaton1, automaton2, automaton3])
    >>> print(events)
    {
        "all": ["a", "b", "c"],
        "controllable": [
          ["a", "b"]
        ],
        "observable": [
          ["a", "b", "c"]
        ],
        "attacker": ["a", "b"]
    }
    """
    if len(automata) < 2:
        raise Exception("Union requires at least two automata in a list as input.")

    # The number of players with different alphabets for observing/controlling
    num_players = len(automata[0]["events"]["controllable"])
    all_events = set()
    cont_events = [set()] * num_players
    obs_events = [set()] * num_players

    # For each automaton, add its events
    for a in automata:
        num_players_local = len(a["events"]["controllable"])
        events = a["events"]
        all_events = all_events.union(set(events["all"]))

        # Add the controllable and observable events from each of the automata
        for i in range(min(num_players, num_players_local)):
            cont_events[i] = cont_events[i].union(set(events["controllable"][i]))
            obs_events[i] = obs_events[i].union(set(events["observable"][i]))

    # Add the events to the result
    return {
        "all": list(all_events),
        # Extract the list of sets into a list of lists
        "controllable": [list(events) for events in cont_events],
        "observable": [list(events) for events in obs_events]
    }


def union_transitions(automata, all_events):
    """Unions the transitions of multiple automata. That is, a transition is
    defined from a state if and only if all automata that have such an event
    defined have a transition defined. This also applies when only one
    automaton has a transition defined.

    This also generates the state space for the unioned automaton.

    Note
    ----
    Any data on bad states or transitions will be not be included in the
    resulting automaton.

    Parameters
    ----------
    automata : array of dictionaries
        Array of all of the automata which should be composed
    all_events : array of strings
        Array of all events that are defined for the automata (even those that
        are only defined in one of them)

    Returns
    -------
    dict
        All of the states and transitions that are defined in a composed union
        of the automata


    Examples
    --------
    In the case of an empty automaton
    >>> print(union_transitions(automaton))
    {
        "states": {
            "all": [],
            "initial": [],
            "bad": [],
            "marked": [[]]
        },
        "transitions": {
            "all": {},
            "bad": {}
        }
    }
    """
    # We'll compute marked states for each of the agents
    num_agents = len(automata[0]["states"]["marked"])
    marked = [[] for x in range(num_agents)]  # The marked states init to empty

    # Get the initial states, mark them as visited (must convert to strings
    # in order to hash them)
    initial_states = state_helper.get_initial(automata)
    initial_strings = [helper.format_state(s) for s in initial_states]

    transitions = {}  # The transitions for the resulting automaton

    # Go through all states systematically
    queue = initial_states.copy()
    visited = set(initial_strings)

    # Keep going through queue until done
    while len(queue) > 0:
        curr = queue.pop(0)
        curr_str = helper.format_state(curr)

        # Add the state to marked, if needed, for each agent
        for agent in range(num_agents):
            if state_helper.check_marked(automata, curr, agent):
                marked[agent].append(curr_str)

        # Go through every event in alphabet
        for event in all_events:
            # Track if the transition does NOT exist in an automaton where it
            # should
            failure = False
            next_state = []
            for i in range(len(automata)):
                # Get the automaton and its state
                a = automata[i]
                prev_state = curr[i]

                # If the state should exist here
                if event in a["events"]["all"]:
                    # Then check if valid
                    transition = helper.format_transition(prev_state, event)
                    if transition in a["transitions"]["all"]:
                        next_state.append(a["transitions"]["all"][transition])
                    else:
                        failure = True  # Not defined
                else:
                    # Then we don't change states
                    next_state.append([prev_state])
            if not failure:
                # Get all possibilities (if nondeterministic)
                next_states = helper.get_states(next_state)
                # Get a string state for use with sets in python
                next_strings = [helper.format_state(nxt) for nxt in next_states]
                # When adding transitions, convert to string format
                transitions[helper.format_transition(helper.format_state(curr), event)] = next_strings
                for i in range(len(next_strings)):
                    if next_strings[i] not in visited:
                        # Then add this state to process next
                        queue.append(next_states[i])
                        visited.add(next_strings[i])
    return {
        "states": {
            "all": list(visited),
            "initial": [helper.format_state(s) for s in initial_states],
            "bad": [],
            "marked": marked
        },
        "transitions": {
            "all": transitions,
            "bad": {}
        }
    }
