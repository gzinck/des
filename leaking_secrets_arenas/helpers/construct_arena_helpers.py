from basic_ops.helpers.string_helpers import *
from arenas.helpers.state_helpers import find_next_state


def add_v2_transitions(automata, state, enabled_events, v2_trans, v2_visited, v1_visited, v1_queue):
    """Adds transitions leaving a given v2 state to v1 and v2 states. If it
    leads to another v2 state, then it adds that v2 state and continues adding
    that state's transitions.

    Parameters
    ----------
    automata : list of dictionaries
        The automata from which the arena is being constructed
    state : list of strings
        The strings representing the current state of each automaton
    enabled_events : set
        The events which are allowed from the current state
    v2_trans : dictionary
        The transitions leaving v2 states
    v2_visited : set
        A set of all visited v2 states (avoids recomputing states)

    Returns
    -------
    None
    """
    # Go through every v2 state we encounter
    queue = [(state, enabled_events)]
    while len(queue) > 0:

        # Turn this info in the queue into string format
        curr = queue.pop(0)
        state = curr[0]
        state_str = format_state(state)
        events = curr[1]
        events_str = format_state_set(events)

        # This is the string representing the current system state
        curr_str = format_state([state_str, events_str])

        # For every possible event that's allowed...
        for event in events:
            nxt = find_next_state(automata, state, event)
            # If this event doesn't actually do anything, ignore
            if nxt == state:
                continue
            # Else, add the transition
            nxt_str = format_state(nxt)
            trans = format_transition(curr_str, event)

            if event in automata[0]["events"]["observable"][0]:
                # Then we go back to a v1
                v2_trans[trans] = [nxt_str]
                if nxt_str not in v1_visited:
                    v1_visited.add(nxt_str)
                    v1_queue.append(nxt)
            else:
                # Then we go to a v2 with same events_str
                nxt_str = format_state([nxt_str, events_str])
                v2_trans[trans] = [nxt_str]
                if nxt_str not in v2_visited:
                    v2_visited.add(nxt_str)
                    queue.append((nxt, events))
