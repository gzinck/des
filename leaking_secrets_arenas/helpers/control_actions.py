from itertools import combinations, chain
import basic_ops.helpers.string_helpers as str_helper
from basic_ops.helpers.determinization_helpers import get_unobservable_reach


def get_valid_control_actions(automaton, state, observable_events):
    """This gets all possible valid control actions for an automaton at a given
    state. That is, it finds all possible actions for the automaton from the
    unobservable reach of the current state and returns its power set as a list
    of tuples. Note than unobservable events will not be included in the sets,
    and all uncontrollable events that may occur at the state will be included
    in every set.

    Notes
    -----
    It does not include the empty set control action

    Parameters
    ----------
    automaton : dictionary
        The automaton for which to get valid control actions
    state : string
        The current state of the automaton
    observable_events : list of strings
        The events which are observable within the automaton

    Returns
    -------
    list of lists
        A list of possible control actions, which are in turn lists of events
        (strings) that are allowed

    Examples
    --------
    >>> print(get_valid_control_actions(automaton, "{q0, q1, q2}"))
    [["a", "u"], ["b", "u"], ["a", "b", "u"]]
    """
    # First, get all the places the automaton could be
    states = get_unobservable_reach(automaton, [state], observable_events)
    # Check for valid controllable events for the controller
    cont_events = automaton["events"]["controllable"][0]
    valid_cevents = []
    for state in states:
        for event in cont_events:
            trans = str_helper.format_transition(state, event)
            if trans in automaton["transitions"]["all"]:
                valid_cevents.append(event)

    # Check for valid uncontrollable events
    all_events = automaton["events"]["all"]
    uncont_events = set(all_events).difference(set(cont_events))
    valid_ucevents = []
    for state in states:
        for event in uncont_events:
            trans = str_helper.format_transition(state, event)
            if trans in automaton["transitions"]["all"]:
                valid_ucevents.append(event)

    # Now that we have all events, get the power set
    all_actions = list(chain.from_iterable(
        combinations(valid_cevents, x) for x in range(0, len(valid_cevents) + 1)
    ))

    # Convert to lists
    events = [list(x) + valid_ucevents for x in all_actions]
    return [x for x in events if len(x) > 0]
