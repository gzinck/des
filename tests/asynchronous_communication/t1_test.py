import unittest
import basic_ops.helpers.convert_to_sets as converter
from structure_validation.automaton_validator import validate
import basic_ops.helpers.string_helpers as helper
import json
from asynchronous_communication.t1_add_comm import add_communication


class TestAddCommAnnotations(unittest.TestCase):
    def setUp(self):
        self.filenames = [
            "tests/asynchronous_communication/test_cases/t1_test.in"
        ]

        self.agent = [
            1
        ]

        self.automata = [{}] * len(self.filenames)
        for i in range(len(self.filenames)):
            with open(self.filenames[i]) as f:
                self.automata[i] = json.load(f)

    def test_add_clock_ticks(self):
        for i in range(len(self.automata)):
            ans = None
            with open(self.filenames[i][:-3] + ".out") as f:
                ans = json.load(f)

            result = add_communication(self.automata[i], self.agent[i])
            validate(result)

            # Print
            # helper.pretty_print(result)

            # Check answer, making sure it's OK if elements not in order
            self.assertEqual(converter.convert_to_sets(result),
                             converter.convert_to_sets(ans))