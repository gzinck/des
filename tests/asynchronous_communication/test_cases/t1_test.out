{
    "events": {
        "all": [
            "a",
            "b",
            "c",
            "tau",
            "?a",
            "?b",
            "?c",
            "!a",
            "!b",
            "!c"
        ],
        "controllable": [
            [
                "a",
                "c"
            ],
            [
                "b",
                "c"
            ]
        ],
        "observable": [
            [
                "a",
                "c"
            ],
            [
                "b",
                "c"
            ]
        ],
        "receive": [
            "?a",
            "?b",
            "?c"
        ],
        "regular": [
            "a",
            "b",
            "c",
            "tau"
        ],
        "send": [
            "!a",
            "!b",
            "!c"
        ]
    },
    "states": {
        "all": [
            "3_0",
            "3_2!",
            "4_1!",
            "6_0?",
            "1_0",
            "1_0?",
            "6_2",
            "2_1",
            "2_1?",
            "6_0",
            "5_1",
            "3_0?",
            "7_1",
            "0",
            "5_1!",
            "3_2",
            "6_2?",
            "4_1",
            "7_1!"
        ],
        "bad": [],
        "initial": [
            "0"
        ],
        "marked": [
            [],
            []
        ]
    },
    "transitions": {
        "all": {
            "0->tau": [
                "1_0"
            ],
            "1_0->?tau": [
                "1_0?"
            ],
            "1_0->a": [
                "2_1"
            ],
            "1_0->b": [
                "5_1"
            ],
            "1_0?->a": [
                "2_1"
            ],
            "1_0?->b": [
                "5_1"
            ],
            "2_1->?a": [
                "2_1?"
            ],
            "2_1->b": [
                "3_2"
            ],
            "2_1?->b": [
                "3_2"
            ],
            "3_0->?tau": [
                "3_0?"
            ],
            "3_0->c": [
                "4_1"
            ],
            "3_0?->c": [
                "4_1"
            ],
            "3_2!->tau": [
                "3_0"
            ],
            "3_2->!b": [
                "3_2!"
            ],
            "3_2->tau": [
                "3_0"
            ],
            "4_1->!c": [
                "4_1!"
            ],
            "5_1!->a": [
                "6_2"
            ],
            "5_1->!b": [
                "5_1!"
            ],
            "5_1->a": [
                "6_2"
            ],
            "6_0->?tau": [
                "6_0?"
            ],
            "6_0->c": [
                "7_1"
            ],
            "6_0?->c": [
                "7_1"
            ],
            "6_2->?a": [
                "6_2?"
            ],
            "6_2->tau": [
                "6_0"
            ],
            "6_2?->tau": [
                "6_0"
            ],
            "7_1->!c": [
                "7_1!"
            ]
        },
        "bad": {},
        "specification": {
            "1_0->a": [
                "2_1"
            ],
            "1_0->b": [
                "5_1"
            ],
            "1_0?->a": [
                "2_1"
            ],
            "1_0?->b": [
                "5_1"
            ],
            "2_1->?a": [
                "2_1?"
            ],
            "2_1->b": [
                "3_2"
            ],
            "2_1?->b": [
                "3_2"
            ],
            "3_0->?tau": [
                "3_0?"
            ],
            "3_0->c": [
                "4_1"
            ],
            "3_0?->c": [
                "4_1"
            ],
            "3_2!->tau": [
                "3_0"
            ],
            "3_2->!b": [
                "3_2!"
            ],
            "3_2->tau": [
                "3_0"
            ],
            "4_1->!c": [
                "4_1!"
            ],
            "5_1!->a": [
                "6_2"
            ],
            "5_1->!b": [
                "5_1!"
            ],
            "5_1->a": [
                "6_2"
            ],
            "6_0->?tau": [
                "6_0?"
            ],
            "6_2->?a": [
                "6_2?"
            ],
            "6_2->tau": [
                "6_0"
            ],
            "6_2?->tau": [
                "6_0"
            ]
        }
    }
}
